spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md

```diff
@@ -488,123 +559,154 @@ NOTE: It is up to individual implementations how they will provide access-contro
+    // counterparty flush status must be FLUSHING or FLUSHINGCOMPLETE
+    abortTransactionUnless(counterpartyFlushStatus == FLUSHING || counterpartyFlushStatus == FLUSHCOMPLETE)
 
-    // verify proofs of counterparty state
-    abortTransactionUnless(verifyChannelState(connection, proofHeight, proofChannel, currentChannel.counterpartyPortIdentifier, currentChannel.counterpartyChannelIdentifier, counterpartyChannel))
+    connection = getConnection(currentChannel.connectionIdentifier)
+    counterpartyHops = getCounterpartyHops(connection)
+
+    // construct counterpartyChannel from existing information and provided
+    // counterpartyUpgradeSequence
+    counterpartyChannel = ChannelEnd{
+        state: TRYUPGRADE,
+        ordering: currentChannel.ordering,
+        counterpartyPortIdentifier: portIdentifier,
+        counterpartyChannelIdentifier: channelIdentifier,
+        connectionHops: counterpartyHops,
+        version: currentChannel.version,
+        sequence: counterpartyUpgradeSequence,
+        flushState: counterpartyFlushStatus,
+    }
 
-    // verify that the counterparty sequence is the same as the current sequence to ensure that the proofs were
-    // retrieved from the current upgrade attempt
-    // since all proofs are retrieved from same proof height, and there can not be multiple upgrade states in the store for a given
-    // channel at the same time
-    sequence = provableStore.get(channelUpgradeSequencePath(portIdentifier, channelIdentifier))
-    abortTransactionUnless(verifyUpgradeSequence(connection, proofHeight, proofUpgradeSequence, currentChannel.counterpartyPortIdentifier,
-    currentChannel.counterpartyChannelIdentifier, sequence))
+    upgrade = provableStore.get(channelUpgradePath(portIdentifier, channelIdentifier))
 
-    // counterparty must be in TRY state
-    if counterpartyChannel.State != TRYUPGRADE {
-        restoreChannel(portIdentifier, channelIdentifier)
-        return
+    // in the crossing hellos case, the versions returned by both on TRY must be the same
+    if currentChannel.state == TRYUPGRADE {
+        if upgrade.fields.version != counterpartyUpgrade.fields.version {
+            restoreChannel(portIdentifier, channelIdentifier)
+        }
     }
 
-    // both channel ends must be mutually compatible.
-    // this means that the ordering must be the same and 
-    // any future introduced fields that must be compatible
-    // should also be checked
-    if counterpartyChannel.ordering != proposedUpgradeChannel.ordering {
-        restoreChannel(portIdentifier, channelIdentifier)
-        return
-    }
+    // prove counterparty and move our own state to ACKUPGRADE and start flushing
+    // upgrade is blocked on this channelEnd from progressing until flush completes on both ends
+    startFlushUpgradeHandshake(portIdentifier, channelIdentifier, upgrade.fields, counterpartyChannel, counterpartyUpgrade, ACKUPGRADE, proofChannel, proofUpgrade, proofHeight)
```

Comment:
Doing this directly became very verbose in the spec and args list, so I keep as-is and simply verify the timeout in the flushing subprotocol.

I do remove the sequence however since we already know this information by this point